pkg/client/introduce/client.go

```diff
@@ -52,77 +60,136 @@ func New(ctx Provider) (*Client, error) {
 		Event:   introduceSvc,
 		service: introduceSvc,
 		store:   store,
+		newUUID: func() string { return uuid.New().String() },
 	}, nil
 }
 
-// SendProposal sends proposal to introducees
+// ProtocolDependency is a helper function that returns the dependency needed for the
+// service to proceed with the protocol. Dependency should be passed to the service through `Continue` function.
+// The function should never return an error, instead of error we provide the callable interface
+// and the state machine will act according to the provided data.
+// Dependency is populated after executing the following functions:
+//  - SendProposal
+//  - SendProposalWithInvitation
+//  - HandleRequest
+//  - HandleRequestWithInvitation
+// usage: e.Continue(c.ProtocolDependency(threadID))
+func (c *Client) ProtocolDependency(thID string) introduce.ActionDependency {
+	opts, err := c.getOptions(thID)
+	if err != nil {
+		opts = &options{}
+
+		logger.Errorf("protocol dependency: %v", err)
+	}
+
+	if opts.Invitation() == nil {
+		opts.Inv = c.defaultInv
+	}
+
+	return opts
+}
+
+// SendProposal sends proposal to introducees (the client does not have a public Invitation)
 func (c *Client) SendProposal(dest1, dest2 *service.Destination) error {
 	return c.sendProposal(options{
-		Destinations: []*service.Destination{dest1, dest2},
+		Dests: []*service.Destination{dest1, dest2},
 	})
 }
 
-// SendProposalWithInvitation sends proposal to introducee
-// executes when the client has a public Invitation
+// SendProposalWithInvitation sends proposal to introducee (the client has a public Invitation)
 func (c *Client) SendProposalWithInvitation(inv *didexchange.Invitation, dest *service.Destination) error {
 	return c.sendProposal(options{
-		Invitation:   inv,
-		Destinations: []*service.Destination{dest},
+		Inv:   inv,
+		Dests: []*service.Destination{dest},
 	})
 }
 
-// SendRequest sends request
+// SendRequest sends a request
 // sending a request means that introducee is willing to share its invitation
 func (c *Client) SendRequest(dest *service.Destination) error {
 	return c.handleOutbound(&introduce.Request{
 		Type: introduce.RequestMsgType,
-		ID:   uuid.New().String(),
+		ID:   c.newUUID(),
+	}, options{
+		Dests: []*service.Destination{dest},
 	})
 }
 
-// HandleRequest executes after receiving a request (the client does not have a public Invitation)
-func (c *Client) HandleRequest(dest1, dest2 *service.Destination) error {
-	return c.sendProposal(options{
-		Destinations: []*service.Destination{dest1, dest2},
+// HandleRequest is a helper function to prepare the right protocol dependency interface
+// It can be executed after receiving a Request action message (the client does not have a public Invitation)
+func (c *Client) HandleRequest(msg service.DIDCommMsg, dest1, dest2 *service.Destination) error {
```

Comment:
Will there be only 2 destinations (dest1 and dest2) ?

Reply:
For now only two destinations. There is some TODO about it https://github.com/hyperledger/aries-rfcs/tree/master/features/0028-introduce#roles

Reply:
ok. I saw the suffix '1' and '2'. hence was just curious.